Sources/Workspace/Workspace.swift

```diff
@@ -2561,168 +2684,208 @@ extension Workspace {
         }
     }
 }
 
-// MARK: - Repository Management
+// MARK: - Package container provider
 
-extension Workspace {
+extension Workspace: PackageContainerProvider {
+    public func getContainer(
+        for package: PackageReference,
+        skipUpdate: Bool,
+        on queue: DispatchQueue,
+        completion: @escaping (Result<PackageContainer, Swift.Error>) -> Void
+    ) {
+        switch package.kind {
+        case .root, .fileSystem:
+            queue.async {
+                // If the container is local, just create and return a local package container.
+                do {
+                    let container = try FileSystemPackageContainer(
+                        package: package,
+                        identityResolver: self.identityResolver,
+                        manifestLoader: self.manifestLoader,
+                        toolsVersionLoader: self.toolsVersionLoader,
+                        currentToolsVersion: self.currentToolsVersion,
+                        fileSystem: self.fileSystem)
+                    completion(.success(container))
+                } catch {
+                    completion(.failure(error))
+                }
+            }
+        case .localSourceControl, .remoteSourceControl:
+            // Resolve the container using the repository manager.
+            do {
+                let repositorySpecifier = try package.makeRepositorySpecifier()
+                repositoryManager.lookup(repository: repositorySpecifier, skipUpdate: skipUpdate, on: queue) { result in
+                    queue.async {
+                        // Create the container wrapper.
+                        let result = result.tryMap { handle -> PackageContainer in
+                            // Open the repository.
+                            //
+                            // FIXME: Do we care about holding this open for the lifetime of the container.
+                            let repository = try handle.open()
+                            return try SourceControlPackageContainer(
+                                package: package,
+                                identityResolver: self.identityResolver,
+                                repositorySpecifier: repositorySpecifier,
+                                repository: repository,
+                                manifestLoader: self.manifestLoader,
+                                toolsVersionLoader: self.toolsVersionLoader,
+                                currentToolsVersion: self.currentToolsVersion
+                            )
+                        }
+                        completion(result)
+                    }
+                }
+            } catch {
+                queue.async {
+                    completion(.failure(error))
+                }
+            }
+        case .registry:
+            fatalError("registry dependencies are supported at this point")
```

Comment:
@mattt this is where you can plug the registry provider

Reply:
```suggestion
            fatalError("registry dependencies are not supported at this point")
```